### PR TITLE
Add delegate/undelegate and block reward core state transition

### PR DESCRIPTION
**Pull request by rlan35**

### CODE REVIEW

File: core/state_transition.go [hunk 1]
```diff
@@ -366,9 +374,77 @@ func (st *StateTransition) applyEditValidatorTx(ev *staking.EditValidator, block
...
+	for i := range wrapper.Delegations {
+		delegation := wrapper.Delegations[i]
+		if bytes.Equal(delegation.DelegatorAddress.Bytes(), delegate.DelegatorAddress.Bytes()) {
+			delegatorExist = true
+			if CanTransfer(stateDB, delegate.DelegatorAddress, delegate.Amount) {
+				delegation.Amount.Add(delegation.Amount, delegate.Amount)
+				err := stateDB.UpdateStakingInfo(wrapper.Validator.Address, wrapper)
+				if err == nil {
+					// TODO: use undelegated token in locking period
+					stateDB.SubBalance(delegate.DelegatorAddress, delegate.Amount)
```
Comment (fxfactorial) [inline core/state_transition.go]: is it ever possible to have a negative balance?

Comment (rlan35) [reply core/state_transition.go]: Good question. It's possible. And the protocol should eliminate those transactions in the very beginning.

File: core/state/statedb.go [hunk 2]
```diff
@@ -726,3 +728,43 @@ func (db *DB) IsValidator(addr common.Address) bool {
 	}
 	return so.IsValidator(db.db)
 }
+
+// AddReward distributes the reward to all the delegators based on stake percentage.
+func (db *DB) AddReward(validator common.Address, reward *big.Int) error {
```
Comment (fxfactorial) [inline core/state/statedb.go]: so if this fails, do I early return as well at call site?

File: core/state_transition.go [hunk 3]
```diff
@@ -366,9 +374,93 @@ func (st *StateTransition) applyEditValidatorTx(ev *staking.EditValidator, block
...
+		delegation := wrapper.Delegations[i]
+		if bytes.Equal(delegation.DelegatorAddress.Bytes(), delegate.DelegatorAddress.Bytes()) {
+			delegatorExist = true
+			totalInUndelegation := delegation.TotalInUndelegation()
+			// If the sum of normal balance and the total amount of tokens in undelegation is greater than the amount to delegate
+			if big.NewInt(0).Add(totalInUndelegation, stateDB.GetBalance(delegate.DelegatorAddress)).Cmp(delegate.Amount) >= 0 {
+				// Firstly use the tokens in undelegation to delegate (redelegate)
+				undelegateAmount := big.NewInt(0).Set(delegate.Amount)
+				// Use the latest undelegated token first as it has the longest remaining locking time.
+				i := len(delegation.Entries) - 1
```
Comment (fxfactorial) [inline core/state_transition.go]: How we know this invariant (latest has longest remaining locking time) is always true? Can sorting list in some way make us keep the invariant?

File: core/state/statedb.go (hunk 2 above)
Comment (rlan35) [reply core/state/statedb.go]: Yes, but the caller should really check the existence of the validator too.

File: core/state_transition.go (hunk 3 above)
Comment (rlan35) [reply core/state_transition.go]: Because we are using this list always as a FIFO queue. It's always ordered by time.

Comment (rlan35) [reply core/state_transition.go]: But added a TODO to sort the list just in case.

**Pull request merged**